operations/kyverno/Kyverno.md

```diff
@@ -0,0 +1,281 @@
+# Kyverno policy management
+
+[Kyverno](https://kyverno.io/) is a policy engine designed specifically for kubernetes. Kyverno allows cluster administrators to manage environment specific configurations independently of workload configurations and enforce configuration best practices for their clusters. Kyverno can be used to scan existing workloads for best practices, or can be used to enforce best practices by blocking or mutating API requests.
+
+It allows you to:
+* Manage policies as Kubernetes resources.
+* Validate, mutate, and generate resource configurations.
+* Select resources based on labels and wildcards.
+* Block non-conformant resources using admission controls, or report policy violations.
+* View policy enforcement as events.
+* Scan existing resources for violations.
+
+Kyverno policies implement the various levels of Kubernetes [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/) for CSM services. The policies are minimally restrictive and enforces the best practices for Pods. The policies make sure the following values are set for workloads (if not present).
+
+```
+securityContext:
+  allowPrivilegeEscalation: false
+  privileged: false
+  runAsUser: 65534
+  runAsNonRoot: true
+  runAsGroup: 65534
+```
+
+Currently, [mutation](#mutation) and [validation](#validation) policies are enforced for the network services such as load balancer and virtualservice.
+
+<a name="mutation"></a>
+## Mutation
+
+Mutation policies are applied in admission controller while creating pods. It mutates the manifest of respective workloads before creating it so that when resource comes up, it will abide by the policy constraints.
+
+######Example:
+
+Step 1: Create a sample policy
+```
+apiVersion: kyverno.io/v1
+kind: Policy
+metadata:
+  name: add-default-securitycontext
+spec:
+  rules:
+  - name: set-container-security-context
+    match:
+      resources:
+        kinds:
+        - Pod
+        selector:
+          matchLabels:
+            app: nginx
+    mutate:
+      patchStrategicMerge:
+        spec:
+          containers:
+          - (name): "*"
+            securityContext:
+              +(allowPrivilegeEscalation): false
+              +(privileged): false
+```
+
+Step 2: Create a simple pod
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    app: nginx
+spec:
+  containers:
+  - name: nginx
+    image: nginx:1.14.2
+    ports:
+    - containerPort: 80
+```
+
+We can list all the policies using following command
+
+```bash
+bash# kubectl get pol -A
+
+NAMESPACE            NAME                        BACKGROUND   ACTION   READY
```

Comment:
Separate the output from the command. For example...

Example output:

```text
NAMESPACE            NAME                        BACKGROUND   ACTION   READY
default              add-default-securitycontext true         audit    true
…
```

Reply:
Done